Refactor: Introduce Validator-Based System for DecentPaymasterV1

adamgall opened the pull request:
Depends on https://github.com/decentdao/decent-contracts/pull/246

Fixes [ENG-817](https://linear.app/decent-labs/issue/ENG-817/refactor-implement-validator-based-system-for-decentpaymasterv1)

**Summary:**

This PR overhauls the `DecentPaymasterV1` to replace its function whitelisting mechanism with a more extensible validator-based system. Instead of directly whitelisting target contracts and function selectors, the paymaster now allows registering dedicated validator contracts for specific functions. This change enhances flexibility and enables more complex validation logic for sponsored UserOperations.

The `BasePaymasterV1` also sees minor updates to its initialization and licensing.

**Key Changes:**

1.  **`DecentPaymasterV1.sol` Overhaul:**
    *   **Inheritance:** Now inherits from `SmartAccountValidationV1` to leverage its UserOperation sender validation.
    *   **Validator System:**
        *   Replaced `_approvedFunctions` mapping with `_functionValidators (address target => mapping(bytes4 selector => address validator))`.
        *   Introduced `setFunctionValidator` to associate a validator contract with a target contract's function.
        *   Added `removeFunctionValidator` to clear a validator.
        *   `getFunctionValidator` now retrieves the registered validator for a function.
    *   **Events & Errors:**
        *   Replaced `FunctionApproved` event with `FunctionValidatorSet` and `FunctionValidatorRemoved`.
        *   Updated error messages to reflect the new system (e.g., `NoValidatorSet`, `ValidationFailed`, `InvalidValidator`).
    *   **Initialization:** The `initialize` function now accepts a `_lightAccountFactory` address and initializes `SmartAccountValidationV1`.
    *   **Validation Logic (`_validatePaymasterUserOp`):**
        1.  Calls `validateUserOp` (from `SmartAccountValidationV1`) to authenticate the UserOperation's sender and extract the target contract and function selector.
        2.  Retrieves the registered `validator` for the target and selector.
        3.  If no validator is found, reverts with `NoValidatorSet`.
        4.  Decodes the inner `callData` from the UserOperation.
        5.  Delegates the final validation to `IFunctionValidator(validator).validateOperation(...)`, passing relevant UserOperation details.
        6.  Reverts with `ValidationFailed` if the validator returns `false`.

2.  **`BasePaymasterV1.sol` Updates:**
    *   SPDX license identifier changed to `GPL-3.0`.
    *   Solidity pragma version updated.
    *   `__BasePaymaster_init` renamed to `__BasePaymasterV1_init`.
    *   Added a default constructor to disable direct initializers.

3.  **New Interface `IFunctionValidator.sol`:**
    *   Defines a standard interface for validator contracts, requiring `supportsInterface(bytes4)` and `validateOperation(address userOpSender, address lightAccountOwner, address target, bytes calldata callData) returns (bool isValid)`.

4.  **Interface `IDecentPaymasterV1.sol` Update:**
    *   Modified to reflect the new validator management functions (`setFunctionValidator`, `removeFunctionValidator`, `getFunctionValidator`).

5.  **New `MockValidator.sol`:**
    *   Added a mock implementation of `IFunctionValidator` for testing purposes, allowing configurable validation outcomes.

6.  **Test Suite `DecentPaymasterV1.test.ts` Rewrite:**
    *   Completely revamped to test the new validator-based architecture.
    *   Includes setup for `MockLightAccount`, `MockGaslessTarget`, `MockLightAccountFactory`, and `MockValidator`.
    *   Proxy deployment and initialization tests updated for the new `_lightAccountFactory` parameter.
    *   "Function Approval" tests replaced with comprehensive "Validator Management" tests.
    *   "Validation" tests now cover scenarios with validators (approval, disapproval, no validator set), and use `MockLightAccount` for constructing `UserOperation.callData`.

**Motivation:**

*   To provide a more powerful and flexible mechanism for determining which UserOperations the `DecentPaymasterV1` should sponsor.
*   To allow for custom, fine-grained validation logic per target function, rather than a simple boolean whitelist.
*   To integrate `SmartAccountValidationV1` for robust sender authentication before applying paymaster-specific validation.
*   To improve the overall design and extensibility of the paymaster system.

dacebt commented on contracts/deployables/account-abstraction/DecentPaymasterV1.sol:
```diff
@@ -58,35 +67,55 @@ contract DecentPaymasterV1 is
-    function whitelistFunctions(
-        address contractAddress,
-        bytes4[] calldata selectors,
-        bool[] calldata approved
+    function setFunctionValidator(
+        address target,
+        bytes4 selector,
+        address validator
```
I don't see in the description, but did have a question, curious around the decision to switch from being able to whitelist multiple functions at a time to setting a single function at a time. Security? curiousity more than anything.

adamgall replied:
Eh no very compelling reason, mostly just simplicity.

Both in the way that our own architecture is built (Safe + Azorius), and with the way that Ethereum UX is headed (7702 delegations), the concept of "batching calls" can happen at a higher level than the target smart contract.

adamgall committed:
Refactor: Introduce Validator-Based System for DecentPaymasterV1

**Summary:**

This PR overhauls the `DecentPaymasterV1` to replace its function whitelisting mechanism with a more extensible validator-based system. Instead of directly whitelisting target contracts and function selectors, the paymaster now allows registering dedicated validator contracts for specific functions. This change enhances flexibility and enables more complex validation logic for sponsored UserOperations.

The `BasePaymasterV1` also sees minor updates to its initialization and licensing.

**Key Changes:**

1.  **`DecentPaymasterV1.sol` Overhaul:**
    *   **Inheritance:** Now inherits from `SmartAccountValidationV1` to leverage its UserOperation sender validation.
    *   **Validator System:**
        *   Replaced `_approvedFunctions` mapping with `_functionValidators (address target => mapping(bytes4 selector => address validator))`.
        *   Introduced `setFunctionValidator` to associate a validator contract with a target contract's function.
        *   Added `removeFunctionValidator` to clear a validator.
        *   `getFunctionValidator` now retrieves the registered validator for a function.
    *   **Events & Errors:**
        *   Replaced `FunctionApproved` event with `FunctionValidatorSet` and `FunctionValidatorRemoved`.
        *   Updated error messages to reflect the new system (e.g., `NoValidatorSet`, `ValidationFailed`, `InvalidValidator`).
    *   **Initialization:** The `initialize` function now accepts a `_lightAccountFactory` address and initializes `SmartAccountValidationV1`.
    *   **Validation Logic (`_validatePaymasterUserOp`):**
        1.  Calls `validateUserOp` (from `SmartAccountValidationV1`) to authenticate the UserOperation's sender and extract the target contract and function selector.
        2.  Retrieves the registered `validator` for the target and selector.
        3.  If no validator is found, reverts with `NoValidatorSet`.
        4.  Decodes the inner `callData` from the UserOperation.
        5.  Delegates the final validation to `IFunctionValidator(validator).validateOperation(...)`, passing relevant UserOperation details.
        6.  Reverts with `ValidationFailed` if the validator returns `false`.

2.  **`BasePaymasterV1.sol` Updates:**
    *   SPDX license identifier changed to `GPL-3.0`.
    *   Solidity pragma version updated.
    *   `__BasePaymaster_init` renamed to `__BasePaymasterV1_init`.
    *   Added a default constructor to disable direct initializers.

3.  **New Interface `IFunctionValidator.sol`:**
    *   Defines a standard interface for validator contracts, requiring `supportsInterface(bytes4)` and `validateOperation(address userOpSender, address lightAccountOwner, address target, bytes calldata callData) returns (bool isValid)`.

4.  **Interface `IDecentPaymasterV1.sol` Update:**
    *   Modified to reflect the new validator management functions (`setFunctionValidator`, `removeFunctionValidator`, `getFunctionValidator`).

5.  **New `MockValidator.sol`:**
    *   Added a mock implementation of `IFunctionValidator` for testing purposes, allowing configurable validation outcomes.

6.  **Test Suite `DecentPaymasterV1.test.ts` Rewrite:**
    *   Completely revamped to test the new validator-based architecture.
    *   Includes setup for `MockLightAccount`, `MockGaslessTarget`, `MockLightAccountFactory`, and `MockValidator`.
    *   Proxy deployment and initialization tests updated for the new `_lightAccountFactory` parameter.
    *   "Function Approval" tests replaced with comprehensive "Validator Management" tests.
    *   "Validation" tests now cover scenarios with validators (approval, disapproval, no validator set), and use `MockLightAccount` for constructing `UserOperation.callData`.

**Motivation:**

*   To provide a more powerful and flexible mechanism for determining which UserOperations the `DecentPaymasterV1` should sponsor.
*   To allow for custom, fine-grained validation logic per target function, rather than a simple boolean whitelist.
*   To integrate `SmartAccountValidationV1` for robust sender authentication before applying paymaster-specific validation.
*   To improve the overall design and extensibility of the paymaster system.
